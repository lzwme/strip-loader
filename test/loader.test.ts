import path from 'path';
import fs from 'fs';
import { compile, getCodeFromBundle, getCompiler, getErrors, getModuleSource, getWarnings } from './helpers';
import type { StripOptions } from '../src/index';

// const isWin = process.platform === 'win32';
const defaultOptions = { debug: false, start: 'devblock:start', end: 'devblock:end' };

describe('strip-loader', () => {
  it('should leave normal files untouched', async () => {
    const testId = 'simple.ts';

    const compiler = getCompiler(testId);
    const stats = await compile(compiler);
    const codeSource = fs.readFileSync(path.resolve(compiler.options.context, testId), 'utf8');
    const moduleSource = getModuleSource(testId, stats) as string;

    expect(codeSource.includes(defaultOptions.start)).toBeFalsy();
    expect(codeSource.includes(defaultOptions.end)).toBeFalsy();
    expect(moduleSource.includes(defaultOptions.start)).toBeFalsy();
    expect(moduleSource.includes(defaultOptions.end)).toBeFalsy();

    const codeFromBundle = getCodeFromBundle(stats, compiler);
    expect(codeFromBundle.map).toBeUndefined();
    expect(codeFromBundle.code).toMatchSnapshot('code');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  }, 10_000);

  it('should trip devblock by default', async () => {
    const testId = 'devblock-default.ts';
    const compiler = getCompiler(testId);
    const stats = await compile(compiler);
    const codeSource = fs.readFileSync(path.resolve(compiler.options.context, testId), 'utf8');
    const moduleSource = getModuleSource(testId, stats) as string;

    expect(codeSource.includes(defaultOptions.start)).toBeTruthy();
    expect(codeSource.includes(defaultOptions.end)).toBeTruthy();
    expect(moduleSource.includes(defaultOptions.start)).toBeFalsy();
    expect(moduleSource.includes(defaultOptions.end)).toBeFalsy();

    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
    expect(moduleSource).toMatchSnapshot();
  }, 10_000);

  it('should not trip devblock when disabled', async () => {
    const testId = 'devblock-default.ts';
    const compiler = getCompiler(testId, { disabled: true });
    const stats = await compile(compiler);
    const codeSource = fs.readFileSync(path.resolve(compiler.options.context, testId), 'utf8');
    const moduleSource = getModuleSource(testId, stats) as string;

    expect(codeSource.includes(defaultOptions.start)).toBeTruthy();
    expect(codeSource.includes(defaultOptions.end)).toBeTruthy();
    expect(moduleSource.includes(defaultOptions.start)).toBeTruthy();
    expect(moduleSource.includes(defaultOptions.end)).toBeTruthy();
  }, 10_000);

  it('should trip devblock by custom options', async () => {
    const customOptions: StripOptions = {
      blocks: [
        {
          start: 'lzwme:debug:start',
          end: 'lzwme:debug:end',
        },
      ],
      isReplaceWithPlaceHolder: false,
      debug: true,
    };
    const testId = 'devblock-custom.ts';
    const compiler = getCompiler(testId, customOptions);
    const stats = await compile(compiler);
    const codeSource = fs.readFileSync(path.resolve(compiler.options.context, testId), 'utf8');
    const moduleSource = getModuleSource(testId, stats) as string;

    expect(codeSource.includes(customOptions.blocks[0].start)).toBeTruthy();
    expect(codeSource.includes(customOptions.blocks[0].end)).toBeTruthy();
    expect(moduleSource.includes(customOptions.blocks[0].start)).toBeFalsy();
    expect(moduleSource.includes(customOptions.blocks[0].end)).toBeFalsy();

    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  }, 10_000);
});
