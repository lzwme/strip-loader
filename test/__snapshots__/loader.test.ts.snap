// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`strip-loader should leave normal files untouched: code 1`] = `undefined`;

exports[`strip-loader should leave normal files untouched: errors 1`] = `[]`;

exports[`strip-loader should leave normal files untouched: warnings 1`] = `[]`;

exports[`strip-loader should trip devblock by custom options: errors 1`] = `[]`;

exports[`strip-loader should trip devblock by custom options: warnings 1`] = `[]`;

exports[`strip-loader should trip devblock by default 1`] = `
"export class AbcDevblockTest {
    constructor() {
        console.log('test for production');
        this.test();
        /* strip-loader */ 
    }
    test() {
        console.log('test for production');
    }
}
"
`;

exports[`strip-loader should trip devblock by default: errors 1`] = `[]`;

exports[`strip-loader should trip devblock by default: warnings 1`] = `[]`;
